cozy/cost_model.py

```diff
@@ -209,6 +280,80 @@ def worst_case_cardinality(e : Exp) -> DominantTerm:
         return DominantTerm.ONE
     return DominantTerm.N
 
+def _maintenance_cost(e : Exp, solver : ModelCachingSolver, op : Op, freebies : [Exp] = []):
+    """Determines the maintenance cost of an expression with regards to its
+    freebies and ops.
+
+    The cost is the result of mutating the expression and getting the storage
+    size of the difference between the mutated expression and the original.
+    """
+    e_prime = mutate(e, op.body)
+    if solver.valid(EEq(e, e_prime)):
+        return ZERO
```

Comment:
Something to try:

In most cases, `alpha_equivalent(e, e_prime)` will be sufficient to detect when `op` does not change the value of `e`. It will also run much much faster.

The downside is that the expression might _appear_ to be affected by `op` when it is not, in which case we would penalize this expression unnecessarily. In most cases I think that is fine.